opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc

```diff
@@ -0,0 +1,220 @@
+
+==== Alarms
+
+The _Kafka Producer_ listens for changes made to the current set of alarms and forwards the resulting alarms to a _Kafka_ topic.
+The records are keyed by alarm reduction key and contain a _GPB_ encoded model of the alarm.
+When an alarm is deleted, a _null_ value is sent with the corresponding reduction key.
+Publishing records in this fashion allows the topic to be used as a link:https://docs.confluent.io/current/streams/concepts.html#ktable[KTable].
+The _Kafka Producer_ will also perform periodic synchronization tasks to ensure that the contents of the Kafka topic reflect the current state of alarms in the _{opennms-product-name}_ database.
+
+By default, all alarms (and subsequent updates) are forwarded to a topic named `alarms`.
+
+The name of the topic used can be configured, and an optional filtering expression can be set to help control which alarms are sent to the topic.
+
+==== Nodes
+
+If an event or alarm being forwarded reference a node, then the corresponding node is also forwarded.
+The records are keyed by "node criteria" (see bellow) and contain a _GPB_ encoded model of the alarm.
+A caching mechanism is in place to help avoid forwarding nodes that have been successfully forwarded, and have not changed since.
+
+The name of the topic used can be configured.
+
+IMPORTANT: The node topic is not intended to include all of the nodes in the system, it only includes records for nodes that relate to events or alarms that have been forwarded.
+
+===== Node Criteria
+
+The _node criteria_ is a string representation of the unique identifier for a given node.
+If the node is associated with a _foreign source (fs)_  and _foreign id (fid)_, the node criteria resulting node criteria will be the name of the _foreign source_, followed by a colon (:) and then the foreign id i.e. (fs:fid).
+If the node is not associated with both a _foreign source_ and _foreign id_, then the node id (database id) will be used.
+
+=== Enabling the Kafka Producer
+
+The _Kafka Producer_ is disabled by default and can be enabled as follows.
+
+First, login to the _Karaf_ shell of your _{opennms-product-name}_ instance and configure the _Kafka_ client settings to point to your _Kafka_ broker.
+See link:https://kafka.apache.org/10/documentation.html#producerconfigs[Producer Configs] for a complete list of available options.
+
+[source]
+----
+$ ssh -p 8101 admin@localhost
+...
+admin@opennms()> config:edit org.opennms.features.kafka.producer.client
+admin@opennms()> config:property-set bootstrap.servers 127.0.0.1:9092
+admin@opennms()> config:update
+----
+
+Next, install the `opennms-kafka-producer` feature from that same shell using:
+
+[source]
+----
+admin@opennms()> feature:install opennms-kafka-producer
+----
+
+In order to ensure that the feature continues to be installed as subsequent restarts, add `opennms-kafka-producer` to the `featuresBoot` property in the `${OPENNMS_HOME}/etc/org.apache.karaf.features.cfg`.
```

Comment:
Just notice and non-critical question. Is there a way to add this to a dedicated file which can be easily updated? As described here, this makes updating and maintaining opennms harder and error prune.

Reply:
Agreed, this same point actually came up earlier today too.

In Minion we can do this in the `featuresBoot.d` folder, we'll be able to do the same in OpenNMS soon (dependent of some refactoring which is being performed as part of HZN-1284).